### PR TITLE
feat(gcp): add service account credentials

### PR DESCRIPTION
**Pull request by pedrooot**
### Description

Add service account key auth method for GCP.

### Checklist

- Are there new checks included in this PR? Yes / No
    - If so, do we need to update permissions for the provider? Please review this carefully.
- [x] Review if the code is being covered by tests.
- [x] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [x] Review if backport is needed.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prowler/providers/gcp/gcp_provider.py [hunk 1]
```diff
@@ -312,25 +321,33 @@ def mutelist(self) -> GCPMutelist:
...
     ) -> tuple:
         """
         Setup the GCP session with the provided credentials file or service account to impersonate
 
         Args:
             credentials_file: str
-            service_account: str
+            service_account: dict
+            gcp_credentials: dict
+            service_account_key: dict
```
Comment (jfagoagas) [inline prowler/providers/gcp/gcp_provider.py]: Please add more details to each argument.

File: prowler/providers/gcp/gcp_provider.py [hunk 2]
```diff
@@ -351,6 +368,24 @@ def setup_session(
                         file=__file__, original_exception=error
                     )
 
+            if service_account_key:
+                logger.info("Using service account key")
```
Comment (jfagoagas) [inline prowler/providers/gcp/gcp_provider.py]:
```suggestion
```

File: prowler/providers/gcp/gcp_provider.py [hunk 3]
```diff
@@ -351,6 +368,24 @@ def setup_session(
                         file=__file__, original_exception=error
                     )
 
+            if service_account_key:
```
Comment (jfagoagas) [inline prowler/providers/gcp/gcp_provider.py:377]: `if` or `elif`?

File: prowler/providers/gcp/gcp_provider.py [hunk 4]
```diff
@@ -388,10 +423,11 @@ def test_connection(
         credentials_file: str = None,
         service_account: str = None,
         raise_on_exception: bool = True,
+        provider_id: Optional[str] = None,
```
Comment (jfagoagas) [inline prowler/providers/gcp/gcp_provider.py:431]: Review if something is calling this function with positional arguments.

File: prowler/providers/gcp/gcp_provider.py [hunk 5]
```diff
@@ -770,6 +788,28 @@ def validate_static_arguments(
             "type": "authorized_user",
         }
 
+    def is_project_matching(self, input_project: str, project_to_match: str) -> bool:
```
Comment (jfagoagas) [inline prowler/providers/gcp/gcp_provider.py]: Can you restore this to their original location?

File: tests/providers/gcp/gcp_provider_test.py [hunk 6]
```diff
@@ -817,6 +803,32 @@ def test_test_connection_with_exception(self):
...
+            side_effect=Exception("Test exception"),
+        ):
+            with pytest.raises(Exception) as e:
+                GcpProvider.test_connection(
+                    service_account_key={"test": "key"},
+                )
+            assert e.type == GCPTestConnectionError
+            assert "Test exception" in e.value.args[0]
+
+    def test_validate_static_arguments(self):
```
Comment (jfagoagas) [inline tests/providers/gcp/gcp_provider_test.py]: Can you restore this to their original location?

**Review by jfagoagas — CHANGES_REQUESTED** [review]

File: tests/providers/gcp/gcp_provider_test.py (hunk 6 above)
Comment (pedrooot) [reply tests/providers/gcp/gcp_provider_test.py]: Done

File: prowler/providers/gcp/gcp_provider.py (hunk 5 above)
Comment (pedrooot) [reply prowler/providers/gcp/gcp_provider.py]: Done

File: prowler/providers/gcp/gcp_provider.py (hunk 4 above)
Comment (pedrooot) [reply prowler/providers/gcp/gcp_provider.py:431]: Done

File: prowler/providers/gcp/gcp_provider.py (hunk 2 above)
Comment (pedrooot) [reply prowler/providers/gcp/gcp_provider.py]: Done

File: prowler/providers/gcp/gcp_provider.py (hunk 3 above)
Comment (pedrooot) [reply prowler/providers/gcp/gcp_provider.py:377]: The elif is not needed, If it matches any statement the return will be triggered.

File: prowler/providers/gcp/gcp_provider.py (hunk 1 above)
Comment (pedrooot) [reply prowler/providers/gcp/gcp_provider.py]: Done

File: prowler/providers/gcp/exceptions/exceptions.py [hunk 7]
```diff
@@ -30,8 +30,8 @@ class GCPBaseException(ProwlerException):
             "message": "Error testing connection to GCP",
             "remediation": "Check the connection and ensure it is properly set up.",
         },
-        (3006, "GCPLoadCredentialsFromDictError"): {
-            "message": "Error loading credentials from dictionary",
+        (3006, "GCPLoadCredentialsFromADCError"): {
+            "message": "Error loading ADC credentials from dictionary associated with ADC",
```
Comment (MrCloudSec) [inline prowler/providers/gcp/exceptions/exceptions.py]:
```suggestion
            "message": "Error loading Application Default Credentials from dictionary",
```

File: prowler/providers/gcp/exceptions/exceptions.py [hunk 8]
```diff
@@ -46,6 +46,10 @@ class GCPBaseException(ProwlerException):
             "message": "Cloud Asset API not used",
             "remediation": "Enable the Cloud Asset API for the project.",
         },
+        (3010, "GCPLoadCredentialsFromServiceAccountKeyError"): {
+            "message": "Error loading credentials from service account key",
```
Comment (MrCloudSec) [inline prowler/providers/gcp/exceptions/exceptions.py]:
```suggestion
            "message": "Error loading Service Account Private Key credentials from dictionary",
```

File: prowler/providers/gcp/exceptions/exceptions.py [hunk 9]
```diff
@@ -30,8 +30,8 @@ class GCPBaseException(ProwlerException):
             "message": "Error testing connection to GCP",
             "remediation": "Check the connection and ensure it is properly set up.",
         },
-        (3006, "GCPLoadCredentialsFromDictError"): {
-            "message": "Error loading credentials from dictionary",
+        (3006, "GCPLoadCredentialsFromADCError"): {
+            "message": "Error loading ADC credentials from dictionary associated with ADC",
             "remediation": "Check the credentials and ensure they are properly set up. client_id, client_secret and refresh_token are required.",
```
Comment (MrCloudSec) [inline prowler/providers/gcp/exceptions/exceptions.py]:
```suggestion
            "remediation": "Check the dictionary and ensure a valid Application Default Credentials are present with client_id, client_secret and refresh_token keys.",
```

File: prowler/providers/gcp/exceptions/exceptions.py [hunk 10]
```diff
@@ -46,6 +46,10 @@ class GCPBaseException(ProwlerException):
             "message": "Cloud Asset API not used",
             "remediation": "Enable the Cloud Asset API for the project.",
         },
+        (3010, "GCPLoadCredentialsFromServiceAccountKeyError"): {
+            "message": "Error loading credentials from service account key",
+            "remediation": "Check the credentials and ensure they are properly set up.",
```
Comment (MrCloudSec) [inline prowler/providers/gcp/exceptions/exceptions.py]:
```suggestion
            "remediation": "Check the dictionary and ensure it contains a Service Account Private Key.",
```

File: prowler/providers/gcp/exceptions/exceptions.py [hunk 11]
```diff
@@ -46,6 +46,10 @@ class GCPBaseException(ProwlerException):
             "message": "Cloud Asset API not used",
             "remediation": "Enable the Cloud Asset API for the project.",
         },
+        (3010, "GCPLoadCredentialsFromServiceAccountKeyError"): {
```
Comment (MrCloudSec) [inline prowler/providers/gcp/exceptions/exceptions.py]:
```suggestion
        (3010, "GCPLoadServiceAccountKeyFromDictError"): {
```

File: prowler/providers/gcp/exceptions/exceptions.py [hunk 12]
```diff
@@ -30,8 +30,8 @@ class GCPBaseException(ProwlerException):
             "message": "Error testing connection to GCP",
             "remediation": "Check the connection and ensure it is properly set up.",
         },
-        (3006, "GCPLoadCredentialsFromDictError"): {
-            "message": "Error loading credentials from dictionary",
+        (3006, "GCPLoadCredentialsFromADCError"): {
```
Comment (MrCloudSec) [inline prowler/providers/gcp/exceptions/exceptions.py]:
```suggestion
        (3006, "GCPLoadADCFromDictError"): {
```

Comment (pedrooot) [reply prowler/providers/gcp/exceptions/exceptions.py]: Done

File: prowler/providers/gcp/exceptions/exceptions.py (hunk 11 above)
Comment (pedrooot) [reply prowler/providers/gcp/exceptions/exceptions.py]: Done

File: prowler/providers/gcp/exceptions/exceptions.py (hunk 10 above)
Comment (pedrooot) [reply prowler/providers/gcp/exceptions/exceptions.py]: Done

File: prowler/providers/gcp/exceptions/exceptions.py (hunk 9 above)
Comment (pedrooot) [reply prowler/providers/gcp/exceptions/exceptions.py]: Done

File: prowler/providers/gcp/exceptions/exceptions.py (hunk 8 above)
Comment (pedrooot) [reply prowler/providers/gcp/exceptions/exceptions.py]: Done

File: prowler/providers/gcp/exceptions/exceptions.py (hunk 7 above)
Comment (pedrooot) [reply prowler/providers/gcp/exceptions/exceptions.py]: Done

**Review by jfagoagas — APPROVED** [review]

**Review by MrCloudSec — APPROVED** [review]

**Pull request merged**